### PR TITLE
Add persistence to zksync payment driver

### PR DESCRIPTION
**Pull request by maaktweluit**
- Extracted database from `gnt` to `base` driver
- Extracted `base/src/cron.rs` from `gnt/src/gnt/sender.rs`
- Implemented Dao in zksync

### CODE REVIEW

File: core/payment-driver/base/src/cron.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+impl Cron {
+    pub fn new(driver: Arc<dyn PaymentDriverCron>) -> Addr<Self> {
+        let me = Self { driver };
+        me.start()
+    }
+
+    fn start_confirmation_job(&mut self, ctx: &mut Context<Self>) {
+        let _ = ctx.run_interval(Duration::from_secs(10), |act, _ctx| {
+            log::debug!("Spawning confirmation job.");
```
Comment (maaktweluit) [inline core/payment-driver/base/src/cron.rs:31]: should be `trace`

File: core/payment-driver/base/src/cron.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            let driver = act.driver.clone();
+            Arbiter::spawn(async move {
+                driver.confirm_payments().await;
+            });
+        });
+    }
+
+    fn start_payment_job(&mut self, ctx: &mut Context<Self>) {
+        let _ = ctx.run_interval(Duration::from_secs(30), |act, _ctx| {
+            log::debug!("Spawning payment job.");
```
Comment (maaktweluit) [inline core/payment-driver/base/src/cron.rs:41]: `trace`

File: core/payment-driver/zksync/src/zksync/wallet.rs [hunk 3]
```diff
@@ -100,6 +100,43 @@ pub async fn make_transfer(details: &PaymentDetails) -> Result<String, GenericEr
...
+    let tx_hash = TxHash::from_str(&tx_hash).unwrap();
+    let tx_info = provider.tx_info(tx_hash).await.unwrap();
+    log::debug!("tx_info: {:?}", tx_info);
+    // Check success
+    match tx_info.success {
+        None => None,
+        Some(result) => {
+            if result {
+                // Only wait for verified blocks on mainnet
+                if *NETWORK != Network::Mainnet || tx_info.is_verified() {
```
Comment (Wiezzel) [inline core/payment-driver/zksync/src/zksync/wallet.rs:115]: Don't wait for verified blocks on mainnet either.

File: core/payment-driver/base/src/bus.rs [hunk 4]
```diff
@@ -64,6 +64,28 @@ pub async fn bind_service<D: PaymentDriver + 'static>(driver: D) {
...
+    let result = service(identity::BUS_ID)
+        .send(message)
+        .await
+        .map_err(GenericError::new)?
+        .map_err(GenericError::new)?;
+    let mut unlocked_list = vec![];
+
+    for node in result {
+        if !node.is_locked {
+            unlocked_list.push(node.node_id);
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/payment-driver/base/src/bus.rs:79]: nice! that looks a lot better, fixed :)

File: core/payment-driver/base/src/dao/transaction.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+impl<'c> AsDao<'c> for TransactionDao<'c> {
+    fn as_dao(pool: &'c PoolType) -> Self {
+        Self { pool }
+    }
+}
+
+impl<'c> TransactionDao<'c> {
+    pub async fn get(&self, tx_id: String) -> DbResult<Option<TransactionEntity>> {
+        do_with_transaction(self.pool, move |conn| {
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/payment-driver/base/src/dao/transaction.rs:33]: Copied from GNT like this.. updated

File: core/payment-driver/base/src/dao/transaction.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            let tx: Option<TransactionEntity> =
+                dsl::transaction.find(tx_id).first(conn).optional()?;
+            match tx {
+                Some(tx) => Ok(Some(tx)),
+                None => Ok(None),
+            }
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/payment-driver/base/src/dao/transaction.rs:39]: Copied from GNT like this.. updated

File: core/payment-driver/base/src/dao/transaction.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                diesel::insert_into(dsl::transaction)
+                    .values(tx)
+                    .execute(conn)?;
+            }
+            Ok(())
+        })
+        .await
+    }
+
+    pub async fn insert_transaction(&self, tx: TransactionEntity) -> DbResult<()> {
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/payment-driver/base/src/dao/transaction.rs:56]: you are right :)

the only thought was that that looks a bit ugly

File: core/payment-driver/base/src/dao/transaction.rs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            Ok(())
+        })
+        .await
+    }
+
+    pub async fn get_used_nonces(&self, address: String) -> DbResult<Vec<String>> {
+        do_with_transaction(self.pool, move |conn| {
+            let nonces: Vec<String> = dsl::transaction
+                .filter(dsl::sender.eq(address))
+                .select(dsl::nonce)
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/payment-driver/base/src/dao/transaction.rs:70]: 👍

File: core/payment-driver/zksync/src/zksync/wallet.rs (hunk 3 above)
Comment (maaktweluit) [reply core/payment-driver/zksync/src/zksync/wallet.rs:115]: 👍

**Review by Wiezzel — APPROVED** [review]
:+1:

**Pull request merged**